pypit/arproc.py

```diff
@@ -836,6 +796,104 @@ def reduce_frame(slf, sciframe, scidx, fitsdict, det, standard=False):
     else: crmask = np.zeros(sciframe.shape)
     # Mask
     slf.update_sci_pixmask(det, crmask, 'CR')
+    return sciframe, rawvarframe, crmask
+
+
+def reduce_echelle(slf, sciframe, scidx, fitsdict, det, standard=False):
+    """ Run standard extraction steps on an echelle frame
+
+    Parameters
+    ----------
+    sciframe : image
+      Bias subtracted image (using arload.load_frame)
+    scidx : int
+      Index of the frame
+    fitsdict : dict
+      Contains relevant information from fits header files
+    det : int
+      Detector index
+    standard : bool, optional
+      Standard star frame?
+    """
+    msgs.work("Multiprocess this algorithm")
+    nspec = sciframe.shape[0]
+    nord = slf._lordloc[det-1].shape[1]
+    # Prepare the frames for tracing and extraction
+    sciframe, rawvarframe, crmask = reduce_prepare(slf, sciframe, scidx, fitsdict, det, standard=standard)
+    bgframe = np.zeros_like(sciframe)
+    if settings.argflag['reduce']['skysub']['perform']:
+        # Identify background pixels, and generate an image of the sky spectrum in each slit
+        for o in range(nord):
+            bgframe += background_subtraction(slf, sciframe, rawvarframe, o, det)
+        modelvarframe = variance_frame(slf, det, sciframe, scidx, fitsdict, skyframe=bgframe)
+    else:
+        modelvarframe = rawvarframe.copy()
+        bgframe = np.zeros_like(sciframe)
+    if not standard:  # Need to save
+        slf._modelvarframe[det - 1] = modelvarframe
+        slf._bgframe[det - 1] = bgframe
+    # Obtain a first estimate of the object trace then
+    # fit the traces and perform a PCA for the refinements
+    trccoeff = np.zeros((settings.argflag['trace']['object']['order'], nord))
+    trcxfit = np.arange(nspec)
+    for o in range(nord):
+        trace, error = artrace.trace_weighted(sciframe-bgframe, slf._lordloc[det-1][:, o], slf._rordloc[det-1][:, o],
+                                              mask=slf._scimask[det-1], wght="flux")
+        msk, trccoeff[:, o] = arutils.robust_polyfit(trcxfit, trace,
+                                                     settings.argflag['trace']['object']['order'],
+                                                     function=settings.argflag['trace']['object']['function'],
+                                                     weights=1.0 / error ** 2, minv=0.0, maxv=nspec-1.0)
+    # Identify the orders to be extrapolated during reconstruction
+    orders = 1.0 + np.arange(nord)
```

Comment:
Since this array is only passed into `func_val`, it will be converted into floats anyway when being multiplied by the fitting coefficients.